File: app/src/main/java/com/ed/redditapp/ui/main/MainModel.java
package com.ed.redditapp.ui.main;

import com.ed.redditapp.lib.api.Post;
import com.ed.redditapp.lib.api.RedditApi;

public class MainModel {
    private final MainPostsAsyncTask mainPostsAsyncTask;
    private final MainSubredditIconUrlAsyncTask subredditIconUrlAsyncTask;

    private Post[] posts;

    public MainModel(RedditApi redditApi) {
        this.mainPostsAsyncTask = new MainPostsAsyncTask(redditApi);
        this.subredditIconUrlAsyncTask = new MainSubredditIconUrlAsyncTask(redditApi);
    }

    public void getMainPagePosts(MainPostsAsyncTask.Callback callback) {
        mainPostsAsyncTask.run((posts) -> {
            this.posts = posts;
            callback.callback(posts);
        });
    }

    public void getMainPageSubredditIcons(MainSubredditIconUrlAsyncTask.Callback callback) {
        subredditIconUrlAsyncTask.run(posts, callback);
    }
}

File: app/src/main/java/com/ed/redditapp/ui/main/MainModel.kt
package com.ed.redditapp.ui.main

import com.ed.redditapp.lib.api.Post

class MainModel(
        val postsAsyncTask: MainPostsAsyncTask,
        val subredditIconUrlAsyncTask: MainSubredditIconUrlAsyncTask) {

    private var posts: Array<Post> = emptyArray()

    fun getMainPagePosts(callback: MainPostsCallback) {
        postsAsyncTask.run(null) { posts ->
            this.posts = posts
            callback(posts)
        }
    }

    fun getNextPosts(after: String, callback: MainPostsCallback) {
        postsAsyncTask.run(after) { posts ->
            this.posts += posts
            callback(posts)
        }
    }

    fun getMainPageSubredditIcons(callback: MainSubredditIconUrlCallback) {
        subredditIconUrlAsyncTask.run(posts, callback)
    }
}